[py2py3] futurize on WMCore/src/python/Utils

mapellidario opened the pull request:
Fixes #9730 

#### Status
ready

#### Description

Run futurize on the modules in the Utils package

The following files have been automatically modified with futurize.
Only some minor manual tweaks have been necessary to make all the
unit tests pass.

	modified:   src/python/Utils/CertTools.py
	modified:   src/python/Utils/EmailAlert.py
	modified:   src/python/Utils/ExtendedUnitTestCase.py
	modified:   src/python/Utils/IteratorTools.py
	modified:   src/python/Utils/MathUtils.py
	modified:   src/python/Utils/MemoryCache.py
	modified:   src/python/Utils/Pipeline.py
	modified:   src/python/Utils/ProcessStats.py
	modified:   src/python/Utils/Signals.py
	modified:   src/python/Utils/Throttled.py
	modified:   src/python/Utils/Timers.py
	modified:   src/python/Utils/Utilities.py

A log of the changes is available at https://cms-dmwm-test.docs.cern.ch/py2py3/porting_log/000_9731_utils/

#### Is it backward compatible (if not, which system it affects?)

yes

#### Related PRs

None

#### External dependencies / deployment changes

It will be necessary to run the unit test in a python3 environment, too

mapellidario committed:
[py2py3] Utils - compatible with both python2 and python3

This commit is related to #9731

The following files have been automatically modified with futurize.

	modified:   src/python/Utils/CertTools.py
	modified:   src/python/Utils/EmailAlert.py
	modified:   src/python/Utils/ExtendedUnitTestCase.py
	modified:   src/python/Utils/IteratorTools.py
	modified:   src/python/Utils/MathUtils.py
	modified:   src/python/Utils/MemoryCache.py
	modified:   src/python/Utils/Pipeline.py
	modified:   src/python/Utils/ProcessStats.py
	modified:   src/python/Utils/Signals.py
	modified:   src/python/Utils/Throttled.py
	modified:   src/python/Utils/Timers.py
	modified:   src/python/Utils/Utilities.py

After applying futurize, the following manual changes
have been applied.

1. Dictionaries

replaced mydict.$operator() with view$operator(mydict)

This is a temporary change and will not be needed anymore
after dropping python2 support. This is used for performance
reasons when running python2 with idioms that are compatibles
with python3.

Resources:
* http://python-future.org/compatible_idioms.html#dictionaries
* http://python-future.org/what_else.html#dict

2. Int and Long

In python3 they have been merged. Int type in py3 is the same as
long type in py2

Resources:
* http://python-future.org/compatible_idioms.html#long-integers
* http://python-future.org/what_else.html#int

3. Bytes and Unicode strings

`from builtins import str` dramatically changed the meaning of
the return of `convertFromUnicodeToStr` function in
src/python/Utils/IteratorTools.py

This function is not used at the moment in dmwm/WMCore,
but it has been updated to use python3-compatible definition
of bytes and unicode strings.

4. Update some older idioms

We use jenkins to check if in the changed files there are some
idioms that are from pre-python 2.6 era. It suggested the change

- if type(expected_obj) != type(actual_obj):
+ if not isinstance(expected_obj, type(actual_obj)):

which comes from 2to3

Resource:
* https://docs.python.org/3/library/2to3.html#2to3fixer-idioms

amaltaro commented:
test this please

amaltaro commented on src/python/Utils/EmailAlert.py:
```diff
@@ -3,11 +3,13 @@
 """
 
 from __future__ import division
+from builtins import str
```
Please group these imports.

amaltaro commented on src/python/Utils/ProcessStats.py:
```diff
@@ -83,7 +83,7 @@ def threadStack():
     id2name = dict([(th.ident, th.name) for th in threading.enumerate()])
     threads = []
-    for tid, stack in sys._current_frames().items():
+    for tid, stack in list(sys._current_frames().items()):
```
can't cast to list here

mapellidario replied:
In this PR I did not change anything by hand. Every change was made automatically by futurize.

Why do you suggest that this can not be done?

The [future doc](http://python-future.org/compatible_idioms.html#dict-keys-values-items-as-a-list) have some examples that look similar to this. Nonetheless, this does not seem to be efficient at all.

As a matter of fact, it seems that future has a way to select the best memory-efficient in both py2 and py3. See
* http://python-future.org/compatible_idioms.html#iterating-through-dict-keys-values-items
* http://python-future.org/what_else.html#dict
* https://stackoverflow.com/a/10190228

amaltaro commented on src/python/Utils/MathUtils.py:
```diff
@@ -15,11 +16,11 @@ def quantize(inputVal, quanta):
         raise ValueError(msg)
 
-    if isinstance(quanta, (basestring, float, long)):
+    if isinstance(quanta, (basestring, float, int)):
```
I'm afraid this might break something. Many of the MySQL queries return long data.

mapellidario replied:
This is strange.

In py3 the int type is like the py2 long type. Futurize let us import the backported py3 int type to py2, so that using int in py2 with futurize import is like using long type in py2 without futurize.

I do not know why this import is not done by default by futirize.

References:
* http://python-future.org/compatible_idioms.html#long-integers
* http://python-future.org/what_else.html#int

amaltaro commented on src/python/Utils/Signals.py:
```diff
@@ -26,7 +26,7 @@ def dumpthreads(isignal, iframe):
     id2name = dict([(th.ident, th.name) for th in threading.enumerate()])
     code = []
-    for tid, stack in sys._current_frames().items():
+    for tid, stack in list(sys._current_frames().items()):
```
same problem casting to a list.

amaltaro commented on src/python/Utils/Utilities.py:
```diff
@@ -18,7 +20,7 @@ def lowerCmsHeaders(headers):
     """
     lheaders = {}
-    for hkey, hval in headers.items(): # perform lower-case
+    for hkey, hval in list(headers.items()): # perform lower-case
```
same problem casting to a list.

amaltaro reviewed (changes requested):
Dario, I have made some comments along the code that will require attention.
We need to see how to deal with the `long` data type as well, given that it's heavily used in MySQL based agents.

Question about the fixer applied to dictionaries. Is that automatically applied by futurize?

mapellidario commented:
test this please

mapellidario commented:
test this please

amaltaro commented on src/python/Utils/CertTools.py:
```diff
@@ -1,4 +1,5 @@
 from __future__ import division, print_function
+from builtins import str
```
It is not clear to me why we need it. Is it to keep the same byte string casting in py2 and py3, thus avoiding unicode strings in python3?

mapellidario replied:
I think futurize automatically added it since it saw that CertTools contained `str()`.

We have to decide what the function should return:

```
[...]
    key = cert = '/tmp/x509up_u' + str(os.getuid())
    if os.path.exists(key):
        return key, cert
```

Right now it is returning two bytes strings. Futurize made the assumption that we want to return two unicode strings and added `from builtins import str`. 

I suggest that even if we want to return bytes, we use `from builtins import bytes` and cast `os.getiud` to bytes, so that the string type will be uniform in python2 and python3.

Some more tests with unicode and bytes strings in py2 are available at https://cms-dmwm-test.docs.cern.ch/py2py3/cheatsheet/strings_future_py2/#unicode-bytes-basestring

amaltaro replied:
What if we change it from:
```
key = cert = '/tmp/x509up_u' + str(os.getuid())
```
to 
```
key = cert = '/tmp/x509up_u{}'.format(os.getuid())
```
?

amaltaro commented on src/python/Utils/ExtendedUnitTestCase.py:
```diff
@@ -23,7 +24,7 @@ def assertContentsEqual(self, expected_obj, actual_obj, msg=None):
 
         def traverse_dict(dictionary):
-            for key, value in dictionary.iteritems():
+            for key, value in iteritems(dictionary):
```
Dario, this documentation:
https://python-future.org/_modules/future/utils.html

seems to suggest to only use `iteritems` if we have python< 2.7. Perhaps we should instead use `viewitems, viewkeys, viewvalues`?

amaltaro commented on src/python/Utils/MathUtils.py:
```diff
@@ -15,11 +17,11 @@ def quantize(inputVal, quanta):
         raise ValueError(msg)
 
-    if isinstance(quanta, (basestring, float, long)):
+    if isinstance(quanta, (basestring, int, float)):
```
no need for `int` here.

amaltaro commented on src/python/Utils/MathUtils.py:
```diff
@@ -15,11 +17,11 @@ def quantize(inputVal, quanta):
     """
     if isinstance(inputVal, basestring):
         inputVal = float(inputVal)
-    elif not isinstance(inputVal, (int, float, long)):
```
removing `long` here scares me. Perhaps we should override the long data type like:
```
if sys.version_info.major == 3:
  long = int
```
? Just an idea, I don't really know what the best way to approach it.

mapellidario replied:
If the plan is to drop python2, having a line that screams `past` was not so appealing :)

The "price" that we have to pay is to get accustomed to the new `int` type that is going to be "long" also in py2, which does not seem terrible. So I took the liberty to choose option 1. I am glad that you agree :)

amaltaro reviewed (changes requested):
Dario, sorry for the very belated review. I have a few questions placed in the code, and likely required changes (e.g. itertools is in several places, but I only made one comment).
Please let me know if anything needs clarification.

mapellidario commented:
I am sorry, described the changes in the commit message, maybe it is better to write the reasoning for the manual changes in the PR comments (i.e. *here* ) instead.


#### PR commit message

Dictionaries: replaced mydict.items() with iteritems(mydict), from future
Resources:
  * http://python-future.org/compatible_idioms.html#dictionaries
  * http://python-future.org/what_else.html#dict
  * https://stackoverflow.com/questions/21783763/in-python-2-what-are-the-performance-trade-offs-for-viewvalues-viewitems-ov
Files
  * Utils/ExtendedUnitTestCase.py
  * Utils/IteratorTools.py
  * Utils/ProcessStats.py
  * Utils/Signals.py
  * Utils/Utilities.py

Int & Long
Resources:
  * http://python-future.org/compatible_idioms.html#long-integers
  * http://python-future.org/what_else.html#int
Files:
  * Utils/MathUtils.py

#### further comments

The approach that futurize has in dealing with ints and longs is why I dropped long in Utils/MathUtils.py

Whether we prefer iteritems or viewitems it is a matter of style. I made a quick research on the topic and found the stackoverflow that I attached to the commit message. The main difference seems that one automatically updates when looping over it, while the other is static. I opted for the one that remains static as a conservative approach, to avoid something like

```
for i in view_of_mylist:
  view_of_mylist.append(something)
```

that could cause not to ever exiting the loop. Maybe this is a silly concern, but I prefer not to loop over dynamic content. As always, the last call is yours, and if we think it is safe for us to loop over the view (which seem to be dynamic), then we can definitely do so.

But in the first link cited in the stackoverflow answer (which is [this](https://stackoverflow.com/a/10190228)) they say that vies may be a bit faster. It is up to you to decide what you prefer.

amaltaro commented:
Dario, thanks for providing this extra documentation. About iteritems vs viewitems, it's a somehow complicated decision.
It looks like `viewitems` provides the same behaviour between py2 and py3 (returning a dict_items), while `iteritems` returns a `dictionary-itemiterator`, which means you can only consume it once. If I understand it correctly, `iteritems` would provide us with a smaller memory footprint in python2, while `viewitems` would have to load the set-like objects in memory at once.

I think having the same behaviour between py2/py3 is important to have as soon as possible. Given that viewitems seems to perform better (even though a test with an empty dict isn't a valid test!), I'm inclined to think that `viewitems` will be a better choice.

mapellidario commented:
Perfect, thanks for the review!
I added the `do not merge` label, will rebase this PR, apply the suggested changes and request for the review again

mapellidario reviewed (commented):
I  added some comments (I am sorry for not having published these earlier yesterday)

mapellidario commented:
test this please

amaltaro commented on src/python/Utils/EmailAlert.py:
```diff
@@ -3,11 +3,12 @@
 
 
-class EmailAlert:
+class EmailAlert(object):
```
If dropping python2 support means more than:
* deleting all the `from __future__ import ...`
* deleting all the `from future import ...`
* deleting all the `from builtins import ...`
so, things that are clearly coming from one of those imports. Then yes, I fully support creating a checklist for python2 support removal.

mapellidario replied:
Well, having `from builtins import object` in python3 is not going to cause any problem, it is only pedantic and verbose. but I can create such a py2 drop checklist!

amaltaro commented on src/python/Utils/IteratorTools.py:
```diff
@@ -30,24 +31,24 @@ def nestedDictUpdate(d, u):
     return d
 
-def convertFromUnicodeToStr(data):
+def convertFromUnicodeToBytes(data):
```
Dario, I believe you did, but let me confirm it. Did you grep the whole code for other places calling/importing this function?

mapellidario replied:
Yep, I did grep all the code and did not find usage of this function.

```
> grep -r "convertFromUnicodeToStr" .
Binary file ./src/python/Utils/IteratorTools.pyc matches
./src/python/Utils/IteratorTools.py:def convertFromUnicodeToStr(data):
./src/python/Utils/IteratorTools.py:        return dict(list(map(convertFromUnicodeToStr, data.iteritems())))
./src/python/Utils/IteratorTools.py:        return type(data)(list(map(convertFromUnicodeToStr, data)))
>
```

amaltaro reviewed (approved):
Changes look good to me, Dario. I made a few comments in the code though.
To be on the safe side, I suggest we only merge it once the final CMSWEB tag has been created.

mapellidario commented:
I squashed the commits into a single commit. All the changes here involved "real" code, no specific "test" code in this PR

amaltaro commented:
Sorry Dario, I overlooked the "Pre-python 2.6 idioms..." from the report file, can you please get that fixed too (and squashed)?

mapellidario commented:
Ok, done! Sorry for not having changed it before. We did not have a discussion about what style we should follow and I changed only what has been deemed necessary. 

The change suggested by the report seems to come from `2to3`, which likes to update some older idioms to more recent alternatives, despite the oldest alternative being still valid. https://docs.python.org/3/library/2to3.html#2to3fixer-idioms

On a same note, would you like me to go thru some of the errors and warning from pylint?

amaltaro commented:
Thanks Dario. About the issue I reported, I think this is the check done in jenkins:
https://github.com/dmwm/WMCore/wiki/Understanding-Jenkins#python-future-code-quality-checks

Regarding refactoring, we can likely leave those to a future opportunity and focus mainly on the python3 support (unless we spot something that might be too fragile).
